der/src/string.rs

```diff
@@ -0,0 +1,105 @@
+//! Common handling for types backed by `String` with enforcement of a
+//! library-level length limitation i.e. `Length::max()`.
+
+use crate::{
+    referenced::OwnedToRef, ByteSlice, DecodeValue, EncodeValue, Header, Length, Reader, Result,
+    StrSlice, Writer,
+};
+use alloc::string::String as StringA;
+use core::str;
+
+/// String newtype which respects the [`Length::max`] limit.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct String {
```

Comment:
I think this is going to be way too confusing name-wise

Reply:
went with `DerString`, again, pretty bad at naming things, and would be happy to take suggestions :)

Reply:
How about:

- `StrSlice` => `StrRef`
- `String` => `StrOwned`
- `ByteSlice` => `BytesRef`
- `Bytes` => `BytesOwned`

...to follow the `Ref`/`Owned` convention